### PR TITLE
Spread the monad!

### PR DESCRIPTION
**Pull request by jeremiedimino**
This PR makes the `gen_rules` callback return a `Memo.Build.t` value, allowing us to use `Memo.Build.t` in many places we couldn't before. This is something we have discussed a few times, and we need it now for porting Jane Street internal build rules to the Dune API. This should also allow us to get rid of `Memo.Sync` and simplify the implementation of `Memo`.

As a result of this change, a few functions in `build_system.mli` get a `Memo.Build.t` such as `eval_pred` or `targets_of`. This causes `Dir_contents` to switch to `Memo.Build`, which in turns contaminate the whole code base. Which is why this PR is huge.

### Added Action_builder.bind

It was difficult to spread the monad without it, and now that we got rid of `external-lib-deps` using bind is not as dramatic. So I introduced it and got rid of `Action_builder.Expert.action_builder`.

But with great powers comes great responsibilities, and it will now be up to us to ensure we write the code in a way that preserve as much parallelism as possible.

### Added Memo.Build.O.(and*)

We don't have that in `Fiber` because it's not clear whether the two sides should be done concurrently or not. In `Memo.Build`, it doesn't matter given that everything is pure.

Since we need `Memo.Build.fork_and_join` everywhere after this PR and the style is heavy, I added `and*`.

### Replaced build_deps passed to Action_builder.Make_exec

It is now only used for `Action_builder.contents` and `Action_builder.lines_of`, so we can use a clearer and simpler function:

```ocaml
val read_file : Path.t -> (Path.t -> 'a) -> 'a Memo.Build.t
```

### Updating the build_system entry point API

I started getting errors saying that an entry point was called inside a memoised function. At this point, calling `do_build` in the middle of the build seems fine, so I reworked a bit all this. I got rid of the `entry_point` function, and instead exposed a single `Build_system.run : 'a Memo.Build.t -> 'a Fiber.t` which takes care of clearing the errors, taking the mutex, etc...

While I was at it, I cleaned up uses of `Memo.Build.t` vs `Fiber.t` in some initialization functions (in particular in `Dune_engine.Main`); now things that do side-effects have type `Fiber.t` and only pure computations have type `Memo.Build.t`. This is important; for instance `Build_system.run` must be called after `Build_system.init` has been called. However, before this PR `Build_system.init` was called inside `Dune_engine.Main.init_build_system` which returned `Memo.Build.t`, which means that it was easy to call it inside the `Build_system.run`, which is clearly a mistake. With the new types, we are forced to call the initialisation functions before entering the memo world.

### Memo.Build.of_fiber --> Memo.Build.of_reproducible_fiber

The previous point made me realise what is required of `Memo.Build.of_fiber`: we need to ensure that the fiber given as argument will always yield the same result. So that we think about this fact when we use it, I renamed it to `Memo.Build.of_reproducible_fiber`.

(I originally chose `unsafe_of_fiber`, but @aalekseyev pointed out that the `unsafe` term is generally related to unsafety in the `Obj.magic` sense. After a discussion with @aalekseyev and @snowleopard we settled for `of_reproducible_fiber`)

### --debug-backtrace now records the location of lazy value

I imported the `Caller_id` module of ppxlib for that. It's really useful for debugging.

### Reworked a bit `dune rules` and the functions it uses

Such as `Build_system.evaluate_rules`. I did this because

I was getting exceptions about `peek_exn` in `Dep.Set.static_deps`, so I got rid of this `peek_exn` rather than try to understand why it was failing.

While I was at it, I updated the code so that `dune rules` expands aliases without building the expansion.

### The language allows more dynamicity

As a natural result of spreading the monad, we are giving more power to users. For instance, we can now use `%{read:..}` in the `deps` field. This is nice, but we should add proper version checks.

I didn't do it in this PR, but opened a ticket to remember to do it: https://github.com/ocaml/dune/issues/4335

### Command.run feels a bit dangerous now

With `Action_builder.bind`, it's easy to write:

```
let* x = f a in
Command.run ... [ ...; Dep y; ... ]
```

and Dune won't discover `y` until `f a` has terminated, thus reducing parallelism. Trying to avoid binds before `Command.run` is a tad difficult, and often result in writing things like this:

```
let x_and_y_and_z = ... in
Command.run (x_and_y_and_z >>| fun x, _, _ -> x) [ (x_and_y_and_z >>| fun _, y, _ -> y); ...]
```

which defeat the purpose of `Command.run`. So I feel like we should reduce the scope of `Command.run` by removing the `Dep` and `Target` constructors.

### Should we have a monad for generating the rules?

At the moment, rules are produced via tracked side effects. But now that everything is monadised, it feels like we could have a monad for that.

### What to do about `Or_exn.t Memo.Build.t`?

We are using `Or_exn.t` in a few places in the code to "delay" errors until rules are executed. After this PR, we have many `Or_exn.t Memo.Build.t` values, so we should probably add some functions to deal with this monad. I didn't do it in this PR.

### API discrepancy between `Action_builder.memoize` and `Memo.Lazy.Async.create`

`Action_builder.memoize` and `Memo.Lazy.Async.create`  share the same purpose, so it seems like we should use the same API. The latter seems easier to reason about as we can use the following rule of thumb:

- a `Memo.Build.t` value should only be used once (to avoid doing several times the same computation)
- only `Memo.Lazy.Async.t` values should be shared

### the "parallel" prefix of "Memo.Build.parallel_*" feels extra

Since everything is pure in the `Memo.Build` monad, everything should be done concurrently.

### Memo.Build is long to type

and:
- it is now used everywhere
- it is more used that just `Memo`

should we swap the two?

### Memo.create_hidden ... Sync is dangerous

If the function type changes from `x -> y` to `x -> y Memo.Build.t`, the system doesn't ask to change `Sync` for `Async` and we end up memoizing a useless value. This happened to me while preparing this PR.

### CODE REVIEW

**Comment by jeremiedimino** [comment]
@rgrinberg why do we need to do `t.errors <- []` in `entry_point*`? `Dune_rules.Main.init_build_system` calls `Build_system.reset` which clears everything anyway.

**Comment by jeremiedimino** [comment]
This PR is now ready for review \o/

**Comment by jeremiedimino** [comment]
Benchmarks:

```
$ ./perf.sh
...
            Current branch Main branch
            ============== =============
          | real 7.705s    real 7.503s
ull build | user 30.996s   user 31.449s
          | sys  16.557s   sys  16.114s

          | real 0.518s    real 0.527s
ero build | user 0.461s    user 0.469s
          | sys  0.057s    sys  0.057s
```

**Comment by rgrinberg** [comment]
> why do we need to do t.errors <- [] in entry_point*? Dune_rules.Main.init_build_system calls Build_system.reset which clears everything anyway.

Then I guess it's not necessary.

File: src/dune_engine/action_builder.ml [hunk 1]
```diff
@@ -5,12 +5,13 @@ module T = struct
...
     | Pure : 'a -> 'a t
     | Map : ('a -> 'b) * 'a t -> 'b t
+    | Bind : 'a t * ('a -> 'b t) -> 'b t
     | Both : 'a t * 'b t -> ('a * 'b) t
     | Seq : unit t * 'b t -> 'b t
     | All : 'a t list -> 'a list t
     | Map2 : ('a -> 'b -> 'c) * 'a t * 'b t -> 'c t
-    | Paths_for_rule : Path.Set.t -> unit t
     | Paths_glob : File_selector.t -> Path.Set.t t
+    | Source_tree : Path.t -> Path.Set.t t
```
Comment (rgrinberg) [inline src/dune_engine/action_builder.ml:14]: Should the argument here be `Path.Source.t`?

Comment (jeremiedimino) [reply src/dune_engine/action_builder.ml:14]: Probably. We need to turn the paths into ones in the current build context, so it would end up being `{ context_name : string; dir : Path.Source.t }`. This seems better, but let's leave that to another PR, for now I kept what we were doing.

**Comment by jeremiedimino** [comment]
I was also wondering why we do `Hooks.End_of_build.once Promotion.finalize` everytime we call `do_build` rather than put it once and for all in `Hooks.End_of_build.always`. We'll do these two changes in another PR.

**Comment by aalekseyev** [comment]
> Added Action_builder.bind

Sounds good!

> Added Memo.Build.O.(and*)

Sounds good!

> Updating the build_system entry point API
I don't fully understand this change. Calling `Build_system.run` from inside `Build_system.run` is probably still problematic, so wouldn't it be good to have a similar check to before?

> --debug-backtrace now records the location of lazy value

Nice!

> While I was at it, I updated the code so that dune rules expands aliases without building the expansion.

Great!

> The language allows more dynamicity

Sounds good too.

> Command.run feels a bit dangerous now ... reducing parallelism.
> So I feel like we should reduce the scope of Command.run by removing the Dep and Target constructors.

I'm a bit confused here. I think I understand the problem, looking at some of the changes you made, but I don't understand the proposal.

> Should we have a monad for generating the rules?

Yeah, I think so!!

> What to do about Or_exn.t Memo.Build.t?
> we should probably add some functions to deal with this monad

Seems reasonable, in particular if it saves some lines of code.

> API discrepancy between Action_builder.memoize and Memo.Lazy.Async.create
> ... we can use the following rule of thumb:
>
> a Memo.Build.t value should only be used once (to avoid doing several times the same computation)
> only Memo.Lazy.Async.t values should be shared

Seems reasonable, but I have a somewhat "meh" feeling about any hard rule here. Duplicating computation is often OK and whatever we do an accidental duplication always remains a possibility. A separate type for `Lazy...t` is nice and useful, but I don't see a problem with sharing e.g. a value `Lazy.force ... : 'a Build.t` if for some reason that's more convenient'.

> the "parallel" prefix of "Memo.Build.parallel_*" feels extra

Agreed.

> Memo.Build is long to type

Maybe simply `Build`, dropping the `Memo` part?
Swapping seems Ok too. (`Memo.t` and `Memo.Function.t`?)

> Memo.create_hidden ... Sync is dangerous

I guess this is related to the "sharing" of Build.t values you mentioned above. Seems unfortunate, but not very different from a simple partial function application. I'm not too worried about it.

File: src/dune_rules/preprocessing.ml [hunk 2]
```diff
@@ -444,15 +442,16 @@ let ppx_driver_and_flags_internal sctx ~loc ~expander ~lib_name ~flags libs =
...
-  let* exe, flags =
-    ppx_driver_and_flags_internal sctx ~loc ~expander ~lib_name ~flags libs
-  in
-  let+ driver =
-    Lib.closure libs ~linking:true >>= Driver.select ~loc:(User_file (loc, pps))
-  in
-  (exe, driver, flags)
+  Action_builder.memo_build
+    (let open Memo.Build.O in
+    let libs = Result.ok_exn (Lib.DB.resolve_pps (Scope.libs scope) pps) in
```
Comment (rgrinberg) [inline src/dune_rules/preprocessing.ml:447]: is it ok to add all these ok_ex here?

**Comment by jeremiedimino** [comment]
> I'm a bit confused here. I think I understand the problem, looking at some of the changes you made, but I don't understand the proposal.

Indeed, let me expand. `Command` provides these two features:

1. you can construct pieces of command lines without knowing in which directory the action will be executed. That seems safe and useful to keep, especially for computing include directories
2. it allows to declare a command line argument and dependency in one go, i.e. instead of having to pass it to both `Action_builder.path` and `Command.run` (or `Action.run`), which is a smaller convenience feature

So what I propose is to change `Command.Args.t` to be:

```ocaml
  type _ t =
    | A : string -> _ t
    | As : string list -> _ t
    | S : 'a t list -> 'a t
    | Concat : string * 'a t list -> 'a t
    | Path : Path.t -> _ t
    | Paths : Path.t list -> _ t
    | Dyn : static t Action_builder.t -> dynamic t
    | Fail : fail -> _ t
    | Expand : expand -> _ t
```

> Seems reasonable, but I have a somewhat "meh" feeling about any hard rule here. Duplicating computation is often OK and whatever we do an accidental duplication always remains a possibility. A separate type for Lazy...t is nice and useful, but I don't see a problem with sharing e.g. a value Lazy.force ... : 'a Build.t if for some reason that's more convenient'.

I think I agree. I'm mostly concerned about shared values that are exposed as `_ Action_builder.t` in the interface, because it is not clear whether it is OK to share them or not. So a weaker rule could be: only values of type `Action_builder.Lazy.t` in mli files can be shared at the call site.

> Maybe simply Build, dropping the Memo part?

Seems good

> I guess this is related to the "sharing" of Build.t values you mentioned above. Seems unfortunate, but not very different from a simple partial function application. I'm not too worried about it.

Agreed. And I guess if we get rid of Sync memos, that will just disappear.

File: src/dune_rules/preprocessing.ml (hunk 2 above)
Comment (jeremiedimino) [reply src/dune_rules/preprocessing.ml:447]: I have been wondering about introducing a `raise : exn -> unit Memo.Build.t` function, and use that exclusively in the rules, for the same purpose

**Comment by aalekseyev** [comment]
>  Command provides these two features:
> 1. `cwd` (more useful)
> 2. deps&targets (less useful)

Huh, I always thought of (2) as being the main point of Command and (1) being an also-nice tack-on feature, so would probably not want to make that change without considering alternatives. (that said, it's definitely worth looking at examples here)

> I'm mostly concerned about shared values that are exposed as _ Action_builder.t in the interface, because it is not clear whether it is OK to share them or not.

Oh, got it. At the minimum me could make a trivial change of introducing `Action_builder.Memoized.t`, but there's probably a cleaner solution in various cases.

File: otherlibs/stdune-unstable/proc.ml [hunk 3]
```diff
@@ -7,7 +7,6 @@ let restore_cwd_and_execve prog argv ~env =
       Unix.create_process_env prog argv env Unix.stdin Unix.stdout Unix.stderr
     in
     match snd (Unix.waitpid [] pid) with
-    | WEXITED 0 -> ()
```
Comment (snowleopard) [inline otherlibs/stdune-unstable/proc.ml:10]: I don't really understand this change: is it related to the goal of this PR, or is it a drive-by fix?

File: src/dune_engine/build_system.mli [hunk 4]
```diff
@@ -85,26 +87,27 @@ val set_vcs : Vcs.t list -> unit Memo.Build.t
 
 (** [prefix_rules t prefix ~f] Runs [f] and adds [prefix] as a dependency to all
     the rules generated by [f] *)
-val prefix_rules : unit Action_builder.t -> f:(unit -> 'a) -> 'a
+val prefix_rules :
```
Comment (snowleopard) [inline src/dune_engine/build_system.mli:90]: Looks like the comments for `prefix_rules` and `eval_pred` got out of date (e.g. there is no `t` anymore). It may be worth fixing this while you're looking at this code.

File: src/dune_engine/rule.mli [hunk 5]
```diff
@@ -50,6 +50,8 @@ module Id : sig
 
   val compare : t -> t -> Ordering.t
 
+  val dummy : t
```
Comment (snowleopard) [inline src/dune_engine/rule.mli:53]: How is this used?

File: test/blackbox-tests/test-cases/foreign-library.t/run.t [hunk 6]
```diff
@@ -91,6 +91,7 @@ Testsuite for the (foreign_library ...) stanza.
   >  (flags -DCONFIG_VALUE=2000)
   >  (names config))
   > EOF
+  $ touch lib/calc.ml
```
Comment (snowleopard) [inline test/blackbox-tests/test-cases/foreign-library.t/run.t:94]: Is this an accidental change?

File: test/blackbox-tests/test-cases/variables-for-artifacts.t/run.t [hunk 7]
```diff
@@ -240,7 +240,7 @@ field of a (rule).
   File "dune", line 3, characters 7-16:
   3 |  (deps %{cmo:x2})
              ^^^^^^^^^
-  Error: %{cmo:..} isn't allowed in this position.
+  Error: Module X2 does not exist.
```
Comment (snowleopard) [inline test/blackbox-tests/test-cases/variables-for-artifacts.t/run.t:243]: Looks like another case, similar to `%{read}`.

**Comment by snowleopard** [comment]
@jeremiedimino Most of the changes you've made look great and uncontroversial to me. To add a couple of points to the above discussions:

* Switching from `Memo.Build.t` to `Build.t` sounds great to me. Adding something like `Build.Or_error.t` would be useful too. 

* I too think that `Dep` and `Target` are pretty useful features of `Command`. It would be sad to lose them. Perhaps, the problem can be solved by adding `Build` versions of `Dep` and `Target`, or otherwise making the Command API more friendly for values that live in the Build monad. In general, there will always be a problem of "sequentiality creep", where uncareful refactoring can introduce regressions (not necessarily in relation to `Command.t`). That may be solvable by testing.

* Looks like you've already made good progress towards getting rid of `Memo.Sync`, so the accidental memoization problem you mentioned should probably be gone soon.

* A rule generation monad would make rule generation a little less mysterious for me. That would be nice.

(All the above should probably be done in separate PRs.)

**Comment by jeremiedimino** [comment]
Yeah, this should all definitely be done in separate PR. This one is big enough already!

FTR, `Dep` and `Target` were indeed the initial motivation for `Command`. But then we started sharing the computation of include directories, and now I can of think of 1 as the more useful feature. In particular, I would use `Command.Args.t Action_builder.Lazy.t` to represent list of include directories, and that would include all the info we need. But happy to iterate more on the design here.

Regarding the changes to `Action_builder` and introducing the rule generation monad, what I have in mind is introducing a functor in `Memo` (or `Build` after the renaming`) which would look like this:

```
module type Data = sig
   type t

   val empty : t
   val union : t -> t -> t

   include Output with type t := t
end

module type Build = sig
   (* interface similar to the current Memo.Build, but augmented with Memo.create, Lazy, etc... *)
end

(* The main monad *)
include Build

module Make(Data : Data) : Build with type 'a t = ('a * Data.t) t
```

and `Action_builder` and the rule generation monad would be two instances of this functor. One with set of facts and one with set of rules. With this functor API, we wouldn't even need to functorise the implementation of memoised functions. The various memoization functions would just be built on top of the central one, by augmenting the `Output` module of memoized functions.

File: otherlibs/stdune-unstable/proc.ml (hunk 3 above)
Comment (jeremiedimino) [reply otherlibs/stdune-unstable/proc.ml:10]: Basically, we were doing a call to `Proc.restore_cwd_and_execve` in a tail position in `bin/exec.ml`. After spreading the monad, this tail position is now expected to have type `unit Fiber.t`. Because the return type of `Proc.restore_cwd_and_execve` is `unit`, it didn't work. The simple fix is to add a `Fiber.return ()` after the call to `Proc.restore_cwd_and_execve`, but this is odd because this point is unreachable. Another way is to make `Proc.restore_cwd_and_execve` return `_`, which is what `Unix.execve` does. To do that, I had to replace this `()` by `exit 0`. 

In any case, this change is good because it aligns more the behaviour of `Proc.restore_cwd_and_execve` between windows and linux.

Comment (jeremiedimino) [reply otherlibs/stdune-unstable/proc.ml:10]: Now that I say this, I realise that we should be calling `sys_exit` rather than `exit`. I updated the code and added a comment.

File: test/blackbox-tests/test-cases/foreign-library.t/run.t (hunk 6 above)
Comment (jeremiedimino) [reply test/blackbox-tests/test-cases/foreign-library.t/run.t:94]: No, it is another case where we started reporting more errors than before. Here, Dune was complaining about `calc.ml` not existing. Since that wasn't the point of the test, I thought it was cluttering the output and added this line to remove the error.

File: test/blackbox-tests/test-cases/variables-for-artifacts.t/run.t (hunk 7 above)
Comment (jeremiedimino) [reply test/blackbox-tests/test-cases/variables-for-artifacts.t/run.t:246]: Indeed, added a TODO30 tag here as well.

File: src/dune_engine/rule.mli (hunk 5 above)
Comment (jeremiedimino) [reply src/dune_engine/rule.mli]: It isn't. It is a leftover I forgot to remove. Thanks for pointing that out. I removed it

File: src/dune_engine/build_system.mli (hunk 4 above)
Comment (jeremiedimino) [reply src/dune_engine/build_system.mli:90]: Indeed, fixed.

File: otherlibs/stdune-unstable/proc.ml (hunk 3 above)
Comment (snowleopard) [reply otherlibs/stdune-unstable/proc.ml:10]: Got it, thanks for the explanation!

File: src/dune_engine/build_system.mli (hunk 4 above)
Comment (snowleopard) [reply src/dune_engine/build_system.mli:90]: There is one more (`eval_pred`) to fix.

Comment (jeremiedimino) [reply src/dune_engine/build_system.mli:90]: Indeed, I updated that as well (and removed the `[`..`]` around the glob argument, which didn't seem useful).

**Comment by snowleopard** [comment]
> Regarding the changes to `Action_builder` and introducing the rule generation monad, what I have in mind is introducing a functor in `Memo` (or `Build` after the renaming`) which would look like this:

I see, that looks similar to the standard [writer monad transformer](https://kseo.github.io/posts/2017-01-21-writer-monad.html), which can be built on top of any inner monad. I guess, since in our case the inner monad is `Memo`, we can incrementalise the computation of the output value (the monoid `Data`), possibly in some interesting ways.

**Review by snowleopard — APPROVED** [review]
That's a big one! Didn't check every line in the diff but made a high-level overview: all looks good.

File: test/blackbox-tests/test-cases/foreign-library.t/run.t (hunk 6 above)
Comment (snowleopard) [reply test/blackbox-tests/test-cases/foreign-library.t/run.t:94]: I see, interesting.

**Comment by jeremiedimino** [comment]
Alright, let's go for it!
